remove deprecated torch.testing functionality

pmeier opened the pull request:
Per title. Deprecation happened on Feb 25, 2022 in c6f1bbc0ac33be0c8ad9956e3fc15e78ddb6cb95, which made it into the 1.12 release. Since it is now 227 days later and the next release will be 1.14, this removal complies with the [BC policy](https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy#minimizing-the-disruption-of-bc-breaking-changes).

For internal users nothing should change as long as they imported everything from `torch.testing._internal`. A quick search through the repo did not turn up anything to the contrary.

pmeier committed:
remove deprecated torch.testing functionality

pmeier committed:
fix

pmeier committed:
lint

pmeier commented:
It seems there is more to do here. I'll convert to draft for now and ping again when CI is green.

pmeier committed:
replace assert_allclose with assert_close

pmeier committed:
fix comment

pmeier committed:
try fix all other failures

pmeier committed:
Merge branch 'master' into remove-deprecated

pmeier committed:
fix more tests

pmeier committed:
fix more caffe/ONNX checks

pmeier committed:
fix fsdp test

pmeier committed:
fix MPS test

pmeier committed:
Merge branch 'master' into remove-deprecated

pmeier committed:
remove legacy import

pmeier committed:
fix caffe tests

pmeier committed:
Merge branch 'master' into remove-deprecated

pmeier committed:
lint

pmeier committed:
fix caffe tests

pmeier commented on caffe2/python/operator_test/_utils.py:
```diff
@@ -0,0 +1,47 @@
+    return max(actual_rtol, expected_rtol), max(actual_atol, expected_atol)
+
+
+# The function below is a faithful replica of the former `torch.testing.assert_allclose`. This is only here,
+# because it is used extensively throughout the tests in this package while needing one feature that
+# the new `torch.testing.assert_close` does not offer: comparison between numpy arrays and torch tensors. See
+# https://github.com/pytorch/pytorch/issues/61844 for the reasoning why this feature was removed.
```
Per comment. This only applies to the tests located in `caffe2/python/operator_test/*`.

pmeier commented on test/distributions/test_distributions.py:
```diff
@@ -1421,7 +1421,7 @@ def ref_log_prob(ref_rate, idx, x, log_prob):
         dist = Poisson(rate_zero)
         dist.log_prob(torch.ones_like(rate_zero)).backward()
-        torch.testing.assert_allclose(rate_zero.grad, torch.inf)
+        self.assertEqual(rate_zero.grad, torch.inf)
```
Comparing scalar tensors to Python scalars is not supported by `torch.testing.assert_close`. I've opted to use `self.assertEqual` here since that still includes this type wrangling for BC.

pmeier commented on test/test_jit_fuser_te.py:
```diff
@@ -2202,7 +2202,7 @@ def test_batch_norm(self):
             trace = torch.jit.trace(fn, args)
             self.assertAllFused(trace.graph_for(*args))
-            torch.testing.assert_allclose(fn(*args), trace(*args))
+            torch.testing.assert_close(fn(*args), trace(*args), equal_nan=True)
```
In `torch.testing.assert_close`, `equal_nan` is `False` by default. Are `NaN`'s actually ok here or did this pass silently before?

mruberry replied:
A comment to follow-up might be nice here, but let's not bother fixing this test in this PR

pmeier commented on test/test_mps.py:
```diff
@@ -60,7 +60,7 @@ def _testRelu(self, np_features, device):
         py_relu_cpu = py_relu.to("cpu")
 
-        torch.testing.assert_allclose(np_relu, py_relu_cpu)
+        self.assertEqual(np_relu, py_relu_cpu)
```
Comparing `numpy.ndarray`'s to `torch.Tensor`'s is not supported in `torch.testing.assert_close`. `self.assertEqual` still allows it.

mruberry replied:
fyi @albanD

albanD replied:
Yes! :)

pmeier commented on test/test_torch.py:
```diff
@@ -2959,10 +2959,9 @@ def test_index_reduce(self, device, dtype, reduce):
-                    if index_noncontig:
-                        # noncontiguous_like fails with RuntimeError: XLA tensors do not have storage
-                        idx = torch.testing.make_non_contiguous(idx)
+                    idx = torch.testing.make_tensor(
```
If we want to create a noncontiguous tensor right away, we can use `torch.testing.make_tensor` directly.

pmeier commented on test/test_torch.py:
```diff
@@ -5588,10 +5587,10 @@ def test_index_add(self):
                         src = torch.randn(num_copy, *other_sizes, device=device)
                         if not src_contig:
-                            src = torch.testing.make_non_contiguous(src)
+                            src = noncontiguous_like(src)
```
Not using `torch.testing.make_tensor` here since `src` samples from a normal distribution and the former would sample from a uniform one. If the intent is to just sample positive as well as negative values, LMK.

pmeier commented on test/test_torch.py:
```diff
@@ -5588,10 +5587,10 @@ def test_index_add(self):
                         idx = torch.randperm(num_dest, dtype=dtype, device=device).narrow(0, 0, num_copy)
                         if not index_contig:
-                            idx = torch.testing.make_non_contiguous(idx)
+                            idx = noncontiguous_like(idx)
```
Same as above but for a random permutation.

pmeier commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -1011,7 +1011,7 @@ def _test_fsdp_parity(
             fsdp_loss = fsdp_loss.cuda()
         fsdp_unsharded_params = get_full_params(fsdp_model)
-        torch.testing.assert_allclose(ref_loss, fsdp_loss)
+        torch.testing.assert_close(ref_loss, fsdp_loss, check_dtype=False)
```
Strict dtype checking is the default for `torch.testing.assert_close`. Are mismatching dtypes ok here?

mruberry replied:
Let's just add a comment for this line so the distributed team can review

@mrshenli

pmeier reviewed (commented):
_The remaining failing CI jobs are unrelated to this PR. They either come from an SEV going on or from periodic tests related to `torchdynamo` / `torchinductor`. Still, can someone import the PR to see if I indeed fixed everything?_

---

The only thing that needed more manual handling was the removal of the undocumented `torch.testing.assert_allclose`. See #61844 for the reasoning behind it. I tried to replace it 1-to-1 with the new-ish (stable since Feb 2022) `torch.testing.assert_close`. That worked most of the time, but sometimes required extra handling since it is not a 100% drop-in replacement. I've highlighted all of these cases below.

pmeier committed:
Merge branch 'master' into remove-deprecated

mruberry commented on caffe2/python/operator_test/_utils.py:
```diff
@@ -0,0 +1,47 @@
+    return max(actual_rtol, expected_rtol), max(actual_atol, expected_atol)
+
+
+# The function below is a faithful replica of the former `torch.testing.assert_allclose`. This is only here,
```
Nice comment -- I would put it at the start of this new file and tweak it to explain why the file exists

mruberry commented on docs/source/fx.rst:
```diff
@@ -1039,7 +1039,7 @@ Miscellanea
 
         x = torch.randn(5, 3)
-        torch.testing.assert_allclose(traced(x), x)
+        torch.testing.assert_close(traced(x), x)
```
Nice docs update

mruberry commented on torch/testing/_internal/common_device_type.py:
```diff
@@ -1314,3 +1311,7 @@ def skipMeta(fn):
     return skipXLAIf(True, "Marked as skipped for XLA")(fn)
+
+
+def get_all_device_types() -> List[str]:
```
We should add a note / file an issue that there are other device types than cpu and cuda (like XLA and MPS) and this should enumerate using the available device type test base classes

mruberry reviewed (commented):
Awesome! Some minor suggestions inline for your review, @pmeier

mruberry commented:
Despite being deprecated for so long there are still many uses of `torch.testing.assert_allclose`:

https://sourcegraph.com/search?q=context:global+torch.testing.assert_allclose+count:100000+lang:python+fork:no&patternType=standard

And this may cause Meta's internal tests to fail. We might need to keep `torch.testing.assert_allclose` available (but the other changes in this PR are great). We could separate the PR into all other changes and the actual removal of `torch.testing.assert_allclose` to de-risk being reverted.

pmeier commented:
I've added a stack at #87969 that splits off each BC breaking change into a different commit. I've copied all the highlights I added here to the stack. @mruberry I also addressed your comments in the stack rather than doing it here.
